File: 138. Copy List with Random Pointer/solution.js
/**
 * Definition for singly-linked list with a random pointer.
 * function RandomListNode(label) {
 *     this.label = label;
 *     this.next = this.random = null;
 * }
 */

/**
 * @param {RandomListNode} head
 * @return {RandomListNode}
 */
var copyRandomList = function(head) {
    var nodes = {};
    var dfs = function(from) {
        if (from === null || from === undefined) return from;
        var node = new RandomListNode(from.label);
        nodes[from.label] = node;
        node.next = dfs(from.next);
        node.random = nodes[from.random];
        return node;
    }
    return dfs(head);
};

File: 138. Copy List with Random Pointer/Solution.java
/**
 * Definition for singly-linked list with a random pointer.
 * class RandomListNode {
 *     int label;
 *     RandomListNode next, random;
 *     RandomListNode(int x) { this.label = x; }
 * };
 */
public class Solution {
    
    public RandomListNode copyRandomList(RandomListNode head) {
        if (head == null) return head;
        HashMap<RandomListNode, RandomListNode> alreadyCopy = new HashMap<>();
        RandomListNode ptr = head;
        while (ptr != null) {
            alreadyCopy.put(ptr, new RandomListNode(ptr.label));
            ptr = ptr.next;
        }
        ptr = head;
        while (ptr != null) {
            RandomListNode copy = alreadyCopy.get(ptr);
            copy.next = ptr.next == null ? null : alreadyCopy.get(ptr.next);
            copy.random = ptr.random == null ? null : alreadyCopy.get(ptr.random);
            ptr = ptr.next;
        }
        return alreadyCopy.get(head);
    }
}